### ISSUE TITLE
New data structure to replace std::unordered_map for component collections

### ISSUE THREAD

**Issue by a-day-old-bagel** [a-day-old-bagel/DualityEngine#21]
Right now the components are held in hash maps for O(1) access time.  Most Entity Components Systems that I've read about use sparse arrays instead.  The problem with a sparse array is that if you have 20 kinds of components, but most entities only use 5 or so of them, that's a _lot_ of wasted memory.  Maybe that's fine.  For all I know, std::unordered_map is implemented with a sparse array anyway.
Moving away from a hash map would be an optimization step, so we don't have to do it right now, or ever if it's not a bottleneck, but if we do have to change it, I thought a sparse array of _pointers_ might be the best way to balance speed and memory at the same time.  That would mean that components themselves are allocated on the heap (not the best option), but that the wasted spaces in the sparse arrays would only be the size of pointers.
Ideas are appreciated.

**Comment by natevm** [comment]
If hashmaps aren't a bottle neck, we should probably leave them alone. Id think they'd be easier to look at in code than a bunch of pointers. 

I'm confused why we have a hashmap of data to begin with... Why isn't everything stored in a component?

**Comment by natevm** [comment]
Edit because I don't read carefully enough. I don't have enough understanding of how this system works to make a proper judgement. I need to pair program with you on building a system before deciding which data structure is best to hold components.

**Comment by a-day-old-bagel** [comment]
I agree that we can wait until it becomes a bottleneck.  I'm just worried that will happen as soon as we start adding or removing components really fast (like if you add a bullet entity each time you fire a gun and then remove each one as it hits something, or something similar).  Of course, we could always design around that, like through manipulating 'active' flags on components instead of actually deleting them, and/or reusing them, but that seems roundabout.  I hope we can implement a system that's fast for addition, retrieval, _and_ removal of components.
